p2p/libp2p/discovery/hostDecorator.go

```diff
@@ -0,0 +1,182 @@
+package discovery
+
+import (
+	"context"
+	"time"
+
+	"github.com/ElrondNetwork/elrond-go/p2p"
+	"github.com/libp2p/go-libp2p-core/connmgr"
+	"github.com/libp2p/go-libp2p-core/event"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	protocol "github.com/libp2p/go-libp2p-core/protocol"
+	peerstore "github.com/libp2p/go-libp2p-peerstore"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+const (
+	hardCPSLimitLow  = 2
+	hardCPSLimitHigh = 100
+)
+
+type hostDecorator struct {
+	h            host.Host
+	acceptedChan chan struct{}
+	resumeChan   chan struct{}
+	timeout      time.Duration
+}
+
+// NewHostDecorator creats a new decorator around an existing host (h) that will olny allow
+// cps connections per second and drop the connection request upon timeout expiration
+func NewHostDecorator(h host.Host, ctx context.Context, cps uint32, timeout time.Duration) (*hostDecorator, error) {
+	if h == nil {
+		return nil, p2p.ErrNilHost
+	}
+
+	if ctx == nil {
+		return nil, p2p.ErrNilContext
+	}
+
+	if timeout < 0 {
+		return nil, p2p.ErrInvalidDurationProvided
+	}
+
+	if cps < hardCPSLimitLow {
+		cps = hardCPSLimitLow
+	}
+
+	if cps > hardCPSLimitHigh {
+		cps = hardCPSLimitHigh
+	}
+
+	ret := &hostDecorator{
+		h:            h,
+		acceptedChan: make(chan struct{}, cps),
+		resumeChan:   make(chan struct{}),
+		timeout:      timeout,
+	}
+
+	go func() {
+		defer close(ret.acceptedChan)
+		for {
+			running := true
+			select {
+			case ret.acceptedChan <- struct{}{}:
+			default:
+				running = false
+			}
+
+			if running {
+				select {
+				case <-time.After(time.Second / time.Duration(cps)):
+				case <-ctx.Done():
+					return
+				}
+			} else {
+				select {
+				case <-ret.resumeChan:
+				case <-ctx.Done():
+					return
+				}
+
+			}
+		}
+	}()
+
+	return ret, nil
+}
+
+// Connect tryes to connect connect to pi if the Connetions Per Second (hd.cps)
```

Comment:
might change "pi" to "pid". Pi refers to something else. Also in the function definition

Reply:
the same is used in the interface definition